File: internal/gandalfcore/policy/policy_test.go
package policy_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/qyinm/gandalf/internal/gandalfcore/policy"
	"github.com/qyinm/gandalf/internal/gandalfcore/types"
)

func TestRestorePolicyFor(t *testing.T) {
	t.Parallel()
	cases := []struct {
		kind types.EvidenceKind
		want types.RestorePolicy
	}{
		{types.KindAgentConfig, types.RestoreFullContent},
		{types.KindMcpServer, types.RestoreStructuredFields},
		{types.KindEnvKey, types.RestoreKeyInventory},
		{types.KindSymlink, types.RestoreNotSupported},
	}
	for _, tc := range cases {
		if got := policy.RestorePolicyFor(tc.kind); got != tc.want {
			t.Fatalf("RestorePolicyFor(%s) = %s, want %s", tc.kind, got, tc.want)
		}
	}
}

func TestSecretLikeKeys(t *testing.T) {
	t.Parallel()
	for _, key := range []string{"OPENAI_API_KEY", "github_token", "client-secret"} {
		if !policy.IsSecretLikeKey(key) {
			t.Fatalf("expected secret-like key %q", key)
		}
	}
	if policy.IsSecretLikeKey("MODEL_NAME") {
		t.Fatal("MODEL_NAME should not be secret-like")
	}
	if policy.CaptureStatusForKey("OPENAI_API_KEY") != "redacted" {
		t.Fatal("expected redacted")
	}
	if policy.CaptureStatusForKey("MODEL_NAME") != "omitted" {
		t.Fatal("expected omitted")
	}
}

func TestRedactStructuredValue(t *testing.T) {
	t.Parallel()
	input := json.RawMessage(`{"command":"npx","api_key":"secret-value","env":{"GITHUB_TOKEN":"abc","MODEL":"gpt-5"}}`)
	redacted, err := policy.RedactStructuredValue(input)
	if err != nil {
		t.Fatalf("redact: %v", err)
	}
	var obj map[string]any
	if err := json.Unmarshal(redacted, &obj); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if obj["api_key"] != "[redacted]" {
		t.Fatalf("api_key = %v", obj["api_key"])
	}
	if obj["command"] != "npx" {
		t.Fatalf("command = %v", obj["command"])
	}
	envKeys, ok := obj["envKeys"].([]any)
	if !ok || len(envKeys) != 2 {
		t.Fatalf("envKeys = %v", obj["envKeys"])
	}
	if _, hasEnv := obj["env"]; hasEnv {
		t.Fatal("env should be removed")
	}
}

func TestIgnoredDirectory(t *testing.T) {
	t.Parallel()
	if !policy.IgnoredDirectory("node_modules") || !policy.IgnoredDirectory(".git") {
		t.Fatal("expected ignored directories")
	}
	if policy.IgnoredDirectory("src") {
		t.Fatal("src should not be ignored")
	}
}

func fixturesDir(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(file), "..", "..", "..", "testdata", "fixtures", "evidence")
}

func TestEvidenceFixtureRoundTrip(t *testing.T) {
	t.Parallel()
	for _, file := range []string{"user-permission-bash.json", "project-permission-bash.json"} {
		raw, err := os.ReadFile(filepath.Join(fixturesDir(t), file))
		if err != nil {
			t.Fatalf("read fixture: %v", err)
		}
		var item types.DiscoveredItem
		if err := json.Unmarshal(raw, &item); err != nil {
			t.Fatalf("deserialize: %v", err)
		}
		serialized, err := json.Marshal(item)
		if err != nil {
			t.Fatalf("serialize: %v", err)
		}
		var roundTrip types.DiscoveredItem
		if err := json.Unmarshal(serialized, &roundTrip); err != nil {
			t.Fatalf("round trip: %v", err)
		}
		if roundTrip.ID != item.ID || roundTrip.Agent != item.Agent || roundTrip.Scope != item.Scope {
			t.Fatalf("round trip mismatch for %s", file)
		}
	}
}

func TestEvidenceFixtureScopes(t *testing.T) {
	t.Parallel()
	dir := fixturesDir(t)
	var user, project types.DiscoveredItem
	if err := json.Unmarshal(mustRead(t, filepath.Join(dir, "user-permission-bash.json")), &user); err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(mustRead(t, filepath.Join(dir, "project-permission-bash.json")), &project); err != nil {
		t.Fatal(err)
	}
	if user.Scope != types.ScopeUser || user.SourcePath != "~/.claude/settings.json" {
		t.Fatalf("user fixture: %+v", user)
	}
	if project.Scope != types.ScopeProject || project.SourcePath != ".claude/settings.json" {
		t.Fatalf("project fixture: %+v", project)
	}
}

func TestInvalidAgentIDDeserializesToUnknown(t *testing.T) {
	t.Parallel()
	raw := json.RawMessage(`{
		"id":"unknown-agent",
		"agent":"not-a-real-agent",
		"kind":"permission",
		"sourcePath":".claude/settings.json",
		"scope":"project",
		"precedence":40,
		"parser":"json",
		"sensitivity":"command_config",
		"contentPolicy":"structured_safe_fields_only",
		"restorePolicy":"structured_fields_only",
		"captureStatus":"captured",
		"confidence":"high"
	}`)
	var item types.DiscoveredItem
	if err := json.Unmarshal(raw, &item); err != nil {
		t.Fatal(err)
	}
	if item.Agent != types.AgentUnknown {
		t.Fatalf("agent = %s", item.Agent)
	}
}

func mustRead(t *testing.T, path string) []byte {
	t.Helper()
	b, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	return b
}

File: crates/gandalf-core/tests/policy_test.rs
use gandalf_core::{
    errors::format_snap_error,
    policy::{
        capture_status_for_key, ignored_directory, is_secret_like_key, redact_structured_value,
        restore_policy_for,
    },
    types::{AgentId, DiscoveredItem, EvidenceKind, EvidenceScope, RestorePolicy},
};
use serde_json::json;
use std::fs;
use std::path::PathBuf;

fn fixtures_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures/evidence")
}

#[test]
fn restore_policy_for_agent_config_is_full_content_supported() {
    assert_eq!(
        restore_policy_for(EvidenceKind::AgentConfig),
        RestorePolicy::FullContentSupported
    );
}

#[test]
fn restore_policy_for_mcp_server_is_structured_fields_only() {
    assert_eq!(
        restore_policy_for(EvidenceKind::McpServer),
        RestorePolicy::StructuredFieldsOnly
    );
}

#[test]
fn restore_policy_for_env_key_is_key_inventory_only() {
    assert_eq!(
        restore_policy_for(EvidenceKind::EnvKey),
        RestorePolicy::KeyInventoryOnly
    );
}

#[test]
fn restore_policy_for_symlink_is_not_supported() {
    assert_eq!(
        restore_policy_for(EvidenceKind::Symlink),
        RestorePolicy::NotSupported
    );
}

#[test]
fn secret_like_keys_trigger_redaction_policy() {
    assert!(is_secret_like_key("OPENAI_API_KEY"));
    assert!(is_secret_like_key("github_token"));
    assert!(is_secret_like_key("client-secret"));
    assert!(!is_secret_like_key("MODEL_NAME"));
    assert_eq!(capture_status_for_key("OPENAI_API_KEY"), "redacted");
    assert_eq!(capture_status_for_key("MODEL_NAME"), "omitted");
}

#[test]
fn redact_structured_value_redacts_secrets_and_expands_env_keys() {
    let input = json!({
        "command": "npx",
        "api_key": "secret-value",
        "env": {
            "GITHUB_TOKEN": "abc",
            "MODEL": "gpt-5"
        }
    });

    let redacted = redact_structured_value(input);
    assert_eq!(redacted["api_key"], "[redacted]");
    assert_eq!(redacted["command"], "npx");
    assert_eq!(redacted["envKeys"], json!(["GITHUB_TOKEN", "MODEL"]));
    assert!(redacted.get("env").is_none());
}

#[test]
fn ignored_directory_matches_ts_policy_set() {
    assert!(ignored_directory("node_modules"));
    assert!(ignored_directory(".git"));
    assert!(!ignored_directory("src"));
}

#[test]
fn evidence_fixture_round_trips_through_serde() {
    for file in ["user-permission-bash.json", "project-permission-bash.json"] {
        let path = fixtures_dir().join(file);
        let raw = fs::read_to_string(&path).expect("fixture should exist");
        let item: DiscoveredItem = serde_json::from_str(&raw).expect("fixture should deserialize");
        let serialized = serde_json::to_string(&item).expect("item should serialize");
        let round_trip: DiscoveredItem =
            serde_json::from_str(&serialized).expect("round trip should deserialize");
        assert_eq!(item, round_trip);
    }
}

#[test]
fn evidence_fixture_scopes_match_user_and_project_boundary() {
    let user: DiscoveredItem = serde_json::from_str(&fs::read_to_string(
        fixtures_dir().join("user-permission-bash.json"),
    )
    .unwrap())
    .unwrap();
    let project: DiscoveredItem = serde_json::from_str(&fs::read_to_string(
        fixtures_dir().join("project-permission-bash.json"),
    )
    .unwrap())
    .unwrap();

    assert_eq!(user.scope, EvidenceScope::User);
    assert_eq!(user.source_path, "~/.claude/settings.json");
    assert_eq!(project.scope, EvidenceScope::Project);
    assert_eq!(project.source_path, ".claude/settings.json");
}

#[test]
fn invalid_agent_id_deserializes_to_unknown() {
    let item: DiscoveredItem = serde_json::from_value(json!({
        "id": "unknown-agent",
        "agent": "not-a-real-agent",
        "kind": "permission",
        "sourcePath": ".claude/settings.json",
        "scope": "project",
        "precedence": 40,
        "parser": "json",
        "sensitivity": "command_config",
        "contentPolicy": "structured_safe_fields_only",
        "restorePolicy": "structured_fields_only",
        "captureStatus": "captured",
        "confidence": "high"
    }))
    .unwrap();

    assert_eq!(item.agent, AgentId::Unknown);
}

#[test]
fn format_snap_error_matches_ts_contract() {
    let output = format_snap_error(&gandalf_core::SnapError {
        code: "GANDALF_PARSE_FAILED".to_string(),
        problem: "Could not parse Codex config.".to_string(),
        cause: "TOML syntax error at line 12.".to_string(),
        fix: "Run `gandalf scan --skip codex` or fix the TOML file.".to_string(),
        path: Some("~/.codex/config.toml".to_string()),
    });

    assert!(output.starts_with("GANDALF_PARSE_FAILED"));
    assert!(output.contains("Problem: Could not parse Codex config."));
    assert!(output.contains("Cause: TOML syntax error at line 12."));
    assert!(output.contains("Fix: Run `gandalf scan --skip codex` or fix the TOML file."));
    assert!(output.contains("Path: ~/.codex/config.toml"));
}